**Import all libraries and modules needed**

In [3]:
import numpy as np
import pandas as pd
import json

from geopy.geocoders import Nominatim

import requests
from pandas.io.json import json_normalize

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

!pip -q install folium
import folium

print('Libraries imported.')

Libraries imported.


**Problem 1: Create a dataframe of the neighborhoods in Toronto from a html file**

In [4]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)

In [5]:
df_toronto = df[0]
df_toronto.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [6]:
#Dropping the rows containing the "Not assigned" cell in the "Borough" column
indexNames = df_toronto[df_toronto['Borough']=='Not assigned'].index
df_toronto.drop(indexNames,axis=0, inplace=True)
df_toronto.reset_index(drop=True, inplace=True)
df_toronto.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [7]:
#Are there any "Not assigned" cell in the "Neighborhood" column in the above dataframe?
cells = df_toronto['Neighborhood'].str.contains('Not assigned')
total = cells.sum()
print('The number of the cells in the Neighborhood column containing "Not assigned":',total)

The number of the cells in the Neighborhood column containing "Not assigned": 0


In [8]:
#Merging rows with the same postal code
df_toronto1 = df_toronto.groupby(['Postal Code','Borough'],as_index=False,sort=False).agg(lambda x: ','.join(x))
df_toronto1.head(11)

Postal Code           Borough                                 Neighborhood
0          M3A        North York                                    Parkwoods
1          M4A        North York                             Victoria Village
2          M5A  Downtown Toronto                    Regent Park, Harbourfront
3          M6A        North York             Lawrence Manor, Lawrence Heights
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5          M9A         Etobicoke      Islington Avenue, Humber Valley Village
6          M1B       Scarborough                               Malvern, Rouge
7          M3B        North York                                    Don Mills
8          M4B         East York              Parkview Hill, Woodbine Gardens
9          M5B  Downtown Toronto                     Garden District, Ryerson
10         M6B        North York                                    Glencairn

In [9]:
df_toronto1.shape

(103, 3)

"df_toronto1" is the name of my final dataframe in Problem 1.  It has 103 rows and 3 columns.

**Problem 2: Create a dataframe of the neighborhoods in Toronto with geographic information**

In [10]:
url_geo = 'https://cocl.us/Geospatial_data/Geospatial_Coordinates.csv'
df_geo = pd.read_csv(url_geo)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [11]:
#Merging two dataframes according to the same Postal Code and creating a new dataframe called df_toronto_geo.
df_toronto_geo = pd.merge(df_toronto1, df_geo, on='Postal Code', how='outer')
df_toronto_geo.head(11)

Postal Code           Borough                                 Neighborhood  \
0          M3A        North York                                    Parkwoods   
1          M4A        North York                             Victoria Village   
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
3          M6A        North York             Lawrence Manor, Lawrence Heights   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5          M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6          M1B       Scarborough                               Malvern, Rouge   
7          M3B        North York                                    Don Mills   
8          M4B         East York              Parkview Hill, Woodbine Gardens   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
10         M6B        North York                                    Glencairn   

     Latitude  Longitude  
0   43.753259 -79.329656  
1   43.725882 -79.315572  
2   43.654260 -79.360636  
3   43.718518 -79.464763  
4   43.662301 -79.389494  
5   43.667856 -79.532242  
6   43.806686 -79.194353  
7   43.745906 -79.352188  
8   43.706397 -79.309937  
9   43.657162 -79.378937  
10  43.709577 -79.445073

In [12]:
df_toronto_geo.shape

(103, 5)

"df_toronto_geo" is the name of my final daraframe in Problem 2.  It has 103 rows and now 5 columns.

**Problem 3: Explore and cluster the neighbborhoods in Toronto**

Get the latitude and longitude values of Toronto.

In [13]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Create a map of Toronto with the neighborhoods superimposed on top.

In [14]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto_geo['Latitude'], df_toronto_geo['Longitude'], df_toronto_geo['Borough'], df_toronto_geo['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Simplify the above map by limitting to the downtown area only.

In [15]:
downtown_data = df_toronto_geo[df_toronto_geo['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Create a map of Downtown Toronto with the neighborhoods superimposed on top.

In [16]:
# create map of Downtown Toronto using latitude and longitude values of Toronto
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

Define Foursquare credentials and version.

In [17]:
# @hidden_cell
CLIENT_ID = 'GPMJJZBX0MHHOAWV52CN3M1QJURTVHBIAZQP1CDNEWGR5G1S'
CLIENT_SECRET = 'QYQC1P0DIYDSHPMXDQOERYWYKDLHYITPTO3KW4QEWADU1UOA'
VERSION = '20200613'
print('The credentials and version were defined.')

The credentials and version were defined.


Expore the first neighborhood(s) (postal code) in the downtown Toronto.

In [18]:
#Getting the names of the first neighborhood(s)
downtown_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [19]:
#Getting the longitude and langitude values of the first neighborhood(s)
neighborhood_latitude = downtown_data.loc[0, 'Latitude']
neighborhood_longitude = downtown_data.loc[0, 'Longitude']

neighborhood_name = downtown_data.loc[0, 'Neighborhood']

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


Get the top 100 venues that are in the first neighborhood(s) within a radius of 500 meters.

In [20]:
radius=500
LIMIT=100
url='https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius,LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=GPMJJZBX0MHHOAWV52CN3M1QJURTVHBIAZQP1CDNEWGR5G1S&client_secret=QYQC1P0DIYDSHPMXDQOERYWYKDLHYITPTO3KW4QEWADU1UOA&ll=43.6542599,-79.3606359&v=20200613&radius=500&limit=100'

In [21]:
#Examining the results
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ee5bc4b237de169927b8dc6'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

Borrow the get_category_type function from the Foursquare lab.

In [22]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Clean the JSON file and structure it to a pandas dataframe.

In [23]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
3  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [24]:
#How many venues are returned by Foursquare?
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

46 venues were returned by Foursquare.


Repeat the same process to all the neighborhoods in dwontown Toronto.

In [25]:
#Creating a function
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [26]:
#Running the above function on each neighborhood(s) and creating a new dataframe called "downton_venues"
downtown_venues = getNearbyVenues(names=downtown_data['Neighborhood'],
                                   latitudes=downtown_data['Latitude'],
                                   longitudes=downtown_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


In [27]:
#Checking the dataframe and its size
print(downtown_venues.shape)
downtown_venues.head()

(1225, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

In [28]:
#Checking how many venues are returned for each neighborhood
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
CN Tower, King and Spadina, Railway Lands, Harb...                     18   
Central Bay Street                                                     65   
Christie                                                               16   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              60   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               95   
Rosedale                                                                4   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         94   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
CN Tower, King and Spadina, Railway Lands, Harb...                      18   
Central Bay Street                                                      65   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               60   
Queen's Park, Ontario Provincial Government                             32   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                                95   
Rosedale                                                                 4   
St. James Town                                                          78   
St. James Town, Cabbagetown                                             48   
Stn A PO Boxes                                                          94   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            56              56   
CN Tower, King and Spadina, Railway Lands, Harb...     18              18   
Central Bay Street                                     65              65   
Christie                                               16              16   
Church and Wellesley                                   78              78   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

In [29]:
#Finding how many uniques categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 213 uniques categories.


Analyze each neighborhood(s)

In [47]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood']=downtown_venues['Neighborhood']

# define a list of column names
cols = downtown_onehot.columns.tolist()
cols

# move the column name to the beggining
cols.insert(0, cols.pop(cols.index('Neighborhood')))
cols

#then use .reindex() function to reorder
downtown_onehot = downtown_onehot.reindex(columns= cols)

downtown_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Trail  Train Station  \
0                    0             0  ...      0              0   
1                    0             0  ...      0              0   
2                    0             0  ...      0              0   
3                    0             0  ...      0              0   
4                    0             0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0         0          0            0              0            0  
1         0          0            0              0            0  
2         0          0            0              0            0  
3         0          0            0              0            0  
4         0          0            0              0            0  

[5 rows x 213 columns]

In [48]:
#Checking the size of the new dataframe
downtown_onehot.shape

(1225, 213)

Group rows by neighborhood(s) and by taking the mean of the frequency of occurrence of each category.

In [49]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
2                                  Central Bay Street           0.000000   
3                                            Christie           0.000000   
4                                Church and Wellesley           0.012821   
5                      Commerce Court, Victoria Hotel           0.000000   
6              First Canadian Place, Underground city           0.000000   
7                            Garden District, Ryerson           0.000000   
8   Harbourfront East, Union Station, Toronto Islands           0.000000   
9           Kensington Market, Chinatown, Grange Park           0.000000   
10        Queen's Park, Ontario Provincial Government           0.000000   
11                          Regent Park, Harbourfront           0.000000   
12                           Richmond, Adelaide, King           0.000000   
13                                           Rosedale           0.000000   
14                                     St. James Town           0.000000   
15                        St. James Town, Cabbagetown           0.000000   
16                                     Stn A PO Boxes           0.000000   
17           Toronto Dominion Centre, Design Exchange           0.000000   
18                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.055556            0.055556      0.055556        0.111111   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000      0.000000        0.000000   
15  0.000000            0.000000      0.000000        0.000000   
16  0.000000            0.000000      0.000000        0.000000   
17  0.000000            0.000000      0.000000        0.000000   
18  0.000000            0.000000      0.000000        0.000000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0          0.000000          0.000000             0.000000      0.000000  ...   
1          0.166667          0.111111             0.000000      0.000000  ...   
2          0.000000          0.000000             0.000000      0.000000  ...   
3          0.000000          0.000000             0.000000      0.000000  ...   
4          0.000000          0.000000             0.012821      0.000000  ...   
5          0.000000          0.000000             0.040000      0.000000  ...   
6          0.000000          0.000000             0.030000      0.000000  ...   
7          0.000000          0.000000             0.000000      0.000000  ...   
8          0.000000          0.000000             0.000000      0.000000  ...   
9          0.000000          0.000000             0.000000      0.000000  ...   
10         0.000000          0.000000             0.000000      0.000000  ...   
11         0.000000          0.000000             0.000000      0.021739  ...   
12         0.000000          0.000000             0.021053      0.000000  ...   
13         0.000000          0.000000             0.000000      0

In [50]:
#Finding a new size
downtown_grouped.shape

(19, 213)

Print each neighborhood(s) along with the top 5 most commmon venues.

In [51]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3  Seafood Restaurant  0.04
4              Bakery  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.17
1    Airport Lounge  0.11
2  Airport Terminal  0.11
3   Harbor / Marina  0.06
4               Bar  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.17
1                 Café  0.08
2   Italian Restaurant  0.06
3  Japanese Restaurant  0.05
4       Sandwich Place  0.05


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3     Baby Store  0.06
4    Candy Store  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1     Sushi Restaurant  0.06
2  Japanese Restaurant  0.05
3           Restaurant  0.04
4  

Create the new dataframe and display the top 10 venues for each neighborhood(s)

In [52]:
#Defining a function that sorts the venues in descending order 
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [53]:
#Creating a new dataframe of the top 10 venues for each neigborhood(s)

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                  Café            Restaurant   
1        Airport Lounge      Airport Terminal              Boutique   
2                  Café    Italian Restaurant        Sandwich Place   
3                  Café                  Park             Nightclub   
4      Sushi Restaurant   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar    Seafood Restaurant                Bakery   
1                 Plane               Airport    Airport Food Court   
2   Japanese Restaurant           Salad Place       Bubble Tea Shop   
3                 Diner    Italian Restaurant            Baby Store   
4               Gay Bar                  Café                   Pub   

  8th Most Common Venue     9th Most Common Venue      10th Most Common Venue  
0           Cheese Shop                 Irish Pub         Japanese Restaurant  
1          Airport Gate       Rental Car Location             Harbor / Marina  
2          Burger Joint          Department Store  Modern European Restaurant  
3            Restaurant               Candy Store                 Coffee Shop  
4           Men's Store  Mediterranean Restaurant                       Hotel

Cluster the neighborhood by running k-means into 5 cluster.

In [54]:
# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 4, 2, 0, 0, 0, 0, 0, 0], dtype=int32)

In [55]:
#Creating a new dataframe including the clusters
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               4           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                Bakery                  Park   
1      Sushi Restaurant           Music Venue              Beer Bar   
2           Coffee Shop                  Café       Bubble Tea Shop   
3                  Café          Cocktail Bar             Gastropub   
4          Cocktail Bar                  Café            Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Restaurant        Breakfast Spot                  Café   
1              Smoothie Shop        Sandwich Place         Burrito Place   
2  Middle Eastern Restaurant   Japanese Restaurant    Italian Restaurant   
3        American Restaurant            Restaurant                   Gym   
4                   Beer Bar    Seafood Restaurant                Bakery   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Theater           Event Space  Performing Arts Venue  
1                  Café                  Park     College Auditorium  
2        Cosmetics Shop              Tea Room              Bookstore  
3                 Hotel    Italian Restaurant         Lingerie Store  
4           Cheese Shop             Irish Pub    Japanese Restaurant

In [56]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters

Example each cluster.

Cluster 1 (red)

In [58]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0                  Café   
12  Downtown Toronto               0                  Café   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                    Pub                Bakery                   Park   
2            Coffee Shop                  Café        Bubble Tea Shop   
3                   Café          Cocktail Bar              Gastropub   
4           Cocktail Bar                  Café             Restaurant   
7             Restaurant                  Café          Deli / Bodega   
8               Aquarium                 Hotel                   Café   
9                  Hotel                  Café             Restaurant   
10                  Café            Restaurant                  Hotel   
11                Bakery                   Bar     Italian Restaurant   
12           Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
15                  Café              Beer Bar             Restaurant   
16            Restaurant                  Café            Pizza Place   
17                  Café            Restaurant                    Gym   
18      Sushi Restaurant   Japanese Restaurant             Restaurant   

            5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                      Restaurant        Breakfast Spot                  Café   
2       Middle Eastern Restaurant   Japanese Restaurant    Italian Restaurant   
3             American Restaurant            Restaurant                   Gym   
4                        Beer Bar    Seafood Restaurant                Bakery   
7                  Clothing Store                   Gym                 Hotel   
8                      Restaurant   Fried Chicken Joint        Scenic Lookout   
9                   Deli / Bodega           Salad Place    Seafood Restaurant   
10                            Gym   American Restaurant         Deli / Bodega   
11            Japanese Restaurant             Bookstore            Restaurant   
12  Vegetarian / Vegan Restaurant          Dessert Shop         Grocery Store   
15            Japanese Restaurant    Italian Restaurant          Cocktail Bar   
16                            Pub                Bakery                  Park   
17                          Hotel   Japanese Restaurant           Salad Place   
18                        Gay Bar                  Café                   Pub   

   8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
0                Theater               Event Space  Performing Arts Venue  
2         Cosmetics Shop                  Tea Room              Bookstore  
3                  Hotel        Italian Restaurant         Lingerie Store  
4            Cheese Shop                 Irish Pub    Japanese Restaurant  
7        Thai Restaurant               Pizza Place                 Bakery  
8    Sporting Goods Shop        Italian Restaurant                Brewery  
9    American Restaurant        Italian Restaurant    Japanese Restaurant  
10   Japanese Restaurant        Italian Restaurant     Seafood Restaurant  
11              Beer Bar                Beer Store         Sandwich Place  
12        

Cluster 2 (purple)

In [60]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground                 Trail         Deli / Bodega   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14   Dumpling Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14               Dog Run   Distribution Center         Discount Store

Cluster 3 (blue)

In [61]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park             Nightclub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6                 Diner    Italian Restaurant            Baby Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6            Restaurant           Candy Store            Coffee Shop

Cluster 4 (green)

In [62]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               3       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13        Airport Lounge      Airport Terminal              Boutique   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13                 Plane               Airport    Airport Food Court   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13          Airport Gate   Rental Car Location        Harbor / Marina

Cluster 5 (orange)

In [63]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1  Downtown Toronto               4           Coffee Shop   
5  Downtown Toronto               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1      Sushi Restaurant           Music Venue              Beer Bar   
5                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1         Smoothie Shop        Sandwich Place         Burrito Place   
5   Japanese Restaurant           Salad Place       Bubble Tea Shop   

  8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
1                  Café                  Park          College Auditorium  
5          Burger Joint      Department Store  Modern European Restaurant